participant/app/routes/$localAgency/recertify/upload.tsx

```diff
@@ -1,54 +1,410 @@
+        noIcon={false}
+        validation={true}
+        type="error"
+        headingLevel="h3"
+      >
+        {errorElements}
+      </Alert>
+    );
+  }
+  return;
+};
 
-import { Button } from "@trussworks/react-uswds";
-import { Form } from "@remix-run/react";
-import { unstable_parseMultipartFormData as parseMultipartFormData } from "@remix-run/server-runtime";
-import type { UploadHandler } from "@remix-run/server-runtime";
-import { useRef } from "react";
-import { useSubmit } from "@remix-run/react";
-import { useTranslation } from "react-i18next";
+export const loader: LoaderFunction = async ({
+  request,
+  params,
+}: {
+  request: Request;
+  params: Params<string>;
+}) => {
+  const { submissionID, headers } = await cookieParser(request, params);
+  const url = new URL(request.url);
+  const removeFileAction = url.searchParams.get("action") == "remove_file";
+  const removeFile = url.searchParams.get("remove");
+  if (removeFileAction && removeFile) {
+    console.log(`⏳ Received request to remove ${removeFile}`);
+    const existingRecord = await findDocument(submissionID, removeFile);
+    if (existingRecord) {
+      await deleteFileFromS3(existingRecord.s3Key);
+      await deleteDocument(submissionID, existingRecord.originalFilename);
+      console.log(
+        `🗑️  Deleted ${existingRecord.originalFilename} from S3 and DB`
+      );
+    } else {
+      console.log(`⚠️  Could not find ${removeFile}`);
+    }
+    // This prevents a remove command from being in the history
+    return redirect(routeRelative(request, "/upload"));
+  }
+  const existingChangesData = (await findSubmissionFormData(
+    submissionID,
+    "changes"
+  )) as ChangesData;
+  if (!existingChangesData) {
+    const returnToChanges = routeRelative(request, "changes");
+    console.log(`No changes data; returning to ${returnToChanges}`);
+    return redirect(returnToChanges);
+  }
+  const proofRequired = determineProof(existingChangesData);
+  if (proofRequired.length == 0) {
+    const skipToContact = routeRelative(request, "contact");
+    console.log(`No proof required; routing to ${skipToContact}`);
+    return redirect(skipToContact);
+  }
+  const previousUploads = await createPreviewData(submissionID);
+  return json(
+    {
+      submissionID: submissionID,
+      proofRequired: proofRequired,
+      maxFileCount: MAX_UPLOAD_FILECOUNT,
+      maxFileSize: MAX_UPLOAD_SIZE_BYTES,
+      previousUploads: previousUploads,
+    },
+    { headers: headers }
+  );
+};
 
-export const action = async ({ request }: { request: Request }) => {
-  const uploadHandler: UploadHandler = async ({
-    name,
-    filename,
-    contentType,
-    data,
-  }) => {
-    if (name !== "documents") {
+export const action = async ({
+  request,
+  params,
+}: {
+  request: Request;
+  params: Params<string>;
+}) => {
+  const { submissionID } = await cookieParser(request, params);
+  const uploadHandler: UploadHandler = async ({ name, filename, data }) => {
```

Comment:
Curious: how would this handle a file input that accepts multiple files? Does it run the handler on each file?

Reply:
Hmmm I think I'm actually getting a little bit turned around about what gets stored in what variable.

Reply:
This is what parses the multipart data; it has no idea how many files will be coming in at once, and it doesn't run serially. 
I believe that it does not care if the file picker is set to 'multiple'; it should work exactly the same.